### DISCUSSION TITLE
How to Integrate Wind Speed Predictions into OpenFAST for Wind Turbine Control?

### DISCUSSION THREAD

**Discussion by mohsamaz** [OpenFAST/openfast#2563] [General]
Hello everyone,

I’m new to OpenFAST and could use some guidance. I’ve developed a wind speed prediction model using Python and AI. The model provides a time series of wind speed predictions based on historical data.
Essentially, I can predict the wind speed for the next few seconds in real-time.

I want to use this predicted data to improve the performance of a wind turbine in OpenFAST.

Here are my questions:

1)How can I provide this predicted wind speed data to OpenFAST?
2)Should this be done through the control system (e.g., in Simulink)?
3)If yes, what type of control system would be appropriate, and how do I start coupling Simulink with OpenFAST?
4)Is there a way to use Python instead of Simulink for this purpose?
Any advice, examples, or resources to help me get started would be greatly appreciated!

Thank you in advance!

**Comment by jjonkman** [comment]
Dear @mohsamaz,

I assume you intend your wind speed prediction model to be used by a custom wind turbine controller.  Can you clarify if you are already using a controller in OpenFAST that you seek to modify?

Here are my responses to your direct questions:

1. By customizing the controller.  I presume it would be easiest to modify an existing controller, such as ROSCO, which is implemented in a Bladed-style DISCON dynamic library rather than in Simulink.  More information on ROSCO is available from: **https://github.com/NREL/ROSCO**.
2. You could use Simulink, but then you'd have to implement your entire controller in Simulink.
3. OpenFAST is already interfaced to MATLAB/Simulink through an S-Function, but you'd have to implement your own control logic in Simulink.
4. Without developing wrapper, e.g., a C-layer compiled as a dynamic library following the Bladed DISCON format that wraps around python, there is no direct way for OpenFAST to call a controller implemented in Python.

Best regards,

**Comment by mohsamaz** [comment]
Dear Dr. Jonkman,  

Thank you for your response.  
I am a complete beginner in OpenFAST and control systems, so I apologize for any inconvenience caused by my questions!  

I am not currently using a controller in OpenFAST or Simulink. My idea is to predict wind speed for the next few seconds and use this prediction to improve system performance. However, I am unsure about which controller to use and how to modify it for this purpose.  

Additionally, I would like to predict rotor speed based on the history of rotor speed data for the next few milliseconds and use this prediction for further improvement.  

Best regards,

**Reply by jjonkman** [reply]
If you are starting fresh, I would recommend extending ROSCO with your additional control logic.  ROSCO is NREL's best reference controller and a great starting point to build from.
